### DISCUSSION TITLE
Metric cache needs to be implemented by ourselves? Docs should be clearer if that's the case

### DISCUSSION THREAD

**Discussion by vesper8** [laravel/nova-issues#4633] [Q&A] [answered]
https://nova.laravel.com/docs/3.0/metrics/defining-metrics.html#caching

I've enabled a 1-hour cache for all my value metrics and I was scratching my head how come the caching doesn't seem to be doing anything.

It appears that while the cache is indeed set, once it's been set it isn't actually loaded anywhere. Are we supposed to implement the load-from-cache ourselves inside our `calculate` method in our metric?

This feels counter-intuitive. Inside the `Metric.php` class, where you create the resolver and then lower in that method you set the cache, couldn't you include a `if cached, return cache` at the top of that method? This is what I would have expected, that if I set the `cacheFor` value then it would automatically cache the results and return the cached content on subsequent loads.

Currently it appears that this is not how it works, and we should use the `getCacheKey` inside our own custom `calculate` method in order to return the cached content. This is a lot of extra code in each metric if that's the case.

Is this really intended to work this way?

Or maybe I'm totally missing something.

Cheers

**Comment by davidhemphill** [comment, marked as answer]
As a sanity check, I created a cached metric inside a new Nova application. It appears caching is working without manually handling the caching internals yourself by just returning a `Carbon` instance inside the `cacheFor` method. 

If you can demonstrate with a reproducible example where metric caching isn't working, we'd love to look at it.

**Reply by vesper8** [reply]
I've done some further testing and it looks like it is indeed caching and not re-entering the `calculate` method, and yet for some reason I feel like it takes a long time just to return cached data, it takes many seconds even though it's loading from cache.. unsure why this is happening. Probably if I was to use something like https://github.com/spatie/laravel-responsecache it would be much faster.
